experiment/src/org/labkey/experiment/api/ExperimentServiceImpl.java

```diff
@@ -8135,6 +8141,98 @@ public List<QueryViewProvider<ExpRun>> getRunOutputsViewProviders()
         return Collections.unmodifiableList(_runOutputsQueryViews);
     }
 
+    @Override
+    public void addDataTypeExclusion(int rowId, DataTypeForExclusion dataType, String excludedContainerId, User user)
+    {
+        Map<String, Object> fields = new HashMap<>();
+        fields.put("DataTypeRowId", rowId);
+        fields.put("DataType", dataType.name());
+        fields.put("ExcludedContainer", excludedContainerId);
+        Table.insert(user, getTinfoDataTypeExclusion(), fields);
+    }
+
+    public void removeDataTypeExclusion(int rowId, DataTypeForExclusion dataType, String excludedContainerId, User user)
+    {
+        SQLFragment sql = new SQLFragment("DELETE FROM  ")
+                .append(getTinfoDataTypeExclusion())
+                .append(" WHERE DataTypeRowId = ? AND DataType = ? AND ExcludedContainer = ?");
+        sql.add(rowId);
+        sql.add(dataType.name());
+        sql.add(excludedContainerId);
+
+        new SqlExecutor(getExpSchema()).execute(sql);
+    }
+
+    @Override
+    public Map<String, Object>[] getContainerDataTypeExclusions(DataTypeForExclusion dataType, String excludedContainerId, Integer dataTypeRowId)
+    {
+        SQLFragment sql = new SQLFragment("SELECT DataTypeRowId, DataType, ExcludedContainer FROM ")
+                .append(getTinfoDataTypeExclusion())
+                .append(" WHERE ");
+        String and = "";
+        if (dataType != null)
+        {
+            sql.append("DataType = ? ");
+            sql.add(dataType.name());
+            and = " AND ";
+        }
+
+        if (!StringUtils.isEmpty(excludedContainerId))
+        {
+            sql.append(and);
+            sql.append("ExcludedContainer = ? ");
+            sql.add(excludedContainerId);
+            and = " AND ";
+        }
+
+        if (dataTypeRowId != null && dataTypeRowId > 0)
+        {
+            sql.append(and);
+            sql.append("DataTypeRowId = ? ");
+            sql.add(dataTypeRowId);
+        }
+
+        return new SqlSelector(getTinfoDataTypeExclusion().getSchema(), sql).getMapArray();
+    }
+
+    public Set<Integer> getContainerDataTypeExclusions(DataTypeForExclusion dataType, String excludedContainerId)
+    {
+        Set<Integer> excludedRowIds = new HashSet<>();
+        Map<String, Object>[] exclusions = getContainerDataTypeExclusions(dataType, excludedContainerId, null);
+        for (Map<String, Object> exclusion : exclusions)
+            excludedRowIds.add((Integer) exclusion.get("DataTypeRowId"));
+
+        return excludedRowIds;
+    }
+
+    @Override
+    public void ensureContainerDataTypeExclusions(DataTypeForExclusion dataType, @Nullable Collection<Integer> excludedDataTypeRowIds, String excludedContainerId, User user)
```

Comment:
```suggestion
    public void ensureContainerDataTypeExclusions(DataTypeForExclusion dataType, @Nullable Collection<Integer> excludedDataTypeRowIds, @Nullable String excludedContainerId, User user)
```

Perhaps this would be better called `updateContainerDataTypeExclusions`?